repeatedly running css3/flexbox/image-percent-max-height.html is failing

obyknovenius opened the pull request:
#### bdfd69d30eda08279f8f3cc1bd7e622b6af34d74
<pre>
repeatedly running css3/flexbox/image-percent-max-height.html is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=267562">https://bugs.webkit.org/show_bug.cgi?id=267562</a>

Reviewed by Alan Baradlay.

Currently, we check if recalculating the preferred logical width
of a flex item is needed only if the style of the flexbox itself
has changed and will cause relayout.

But flexbox relayout can be caused by other reasons e.g. when its
size is relative to its parent and the parent size has changed.

This patch moves the check from `RenderFlexibleBox::styleDidChange()`
to `RenderFlexibleBox::constructFlexItem()`.

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange):
(WebCore::RenderFlexibleBox::constructFlexItem):

Canonical link: <a href="https://commits.webkit.org/278461@main">https://commits.webkit.org/278461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bba381801cfaf5d3d478d62998052d63f1be1eb0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53113 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/547 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35197 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/114 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40687 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26712 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42903 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21811 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24148 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/115 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8240 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46165 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54694 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24964 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/109 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48073 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26221 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43079 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47111 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11092 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27083 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1709,6 +1706,9 @@ FlexItem RenderFlexibleBox::constructFlexItem(RenderBox& child, bool relayoutChi
     if (childHadLayout && child.hasTrimmedMargin(std::optional<MarginTrimType> { }))
         child.clearTrimmedMarginsMarkings();
     
+    if (child.needsPreferredWidthsRecalculation())
+        child.setPreferredLogicalWidthsDirty(true, MarkingBehavior::MarkOnlyThis);
+
```
I think it'd be better to have a dedicated function to run invalidation on flex items in a similar way we do it in block layout, see `RenderBlock::updateBlockChildDirtyBitsBeforeLayout` (as opposed to put it in a seemingly random function)

obyknovenius replied:
There is `updateFlexItemDirtyBitsBeforeLayout()` already. But it is called in `RenderFlexibleBox::layoutAndPlaceChildren()` which is too late because flex items have already been constructed by this time.

obyknovenius replied:
@alanbaradlay What do you think? Should I land it as is?

alanbaradlay replied:
sure.

obyknovenius committed:
repeatedly running css3/flexbox/image-percent-max-height.html is failing
https://bugs.webkit.org/show_bug.cgi?id=267562

Reviewed by Alan Baradlay.

Currently, we check if recalculating the preferred logical width
of a flex item is needed only if the style of the flexbox itself
has changed and will cause relayout.

But flexbox relayout can be caused by other reasons e.g. when its
size is relative to its parent and the parent size has changed.

This patch moves the check from `RenderFlexibleBox::styleDidChange()`
to `RenderFlexibleBox::constructFlexItem()`.

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::styleDidChange):
(WebCore::RenderFlexibleBox::constructFlexItem):

Canonical link: https://commits.webkit.org/278461@main
